PKI/TLS improvements (#13622)
cockroachdb/docs: v22.1/manage-certs-vault.md

                    Public-Key: (2048 bit)
          ...
    ~~~

    
### Create PKI roles and issue certificates for the nodes

In Vault, a PKI role is a template for a certificate.

1.  Create a node role for each node. The role will be used to generate certifates for the corresponding node.

    Each certificate is tailored to the node:
    - The extended key usages attribute `ext_key_usage` must include both server and client auth usages; this is because nodes must frequently initiate requests to other nodes in order to maintain cluster synchrony and load-balance work.
    - The Subject Alternative Name (SAN) - IP addresses field contains:
      - the IP address of the node on the internal network of your GCP project (so the node can serve at that address  locally, to other nodes).
      - the IP address of your cluster on the external, public internet (so the node can serve at that address publicly, to application servers).

    {{site.data.alerts.callout_info}}
    Note that certificate attributes must be provided with Vault's specific parameter syntax, which is documented here: 
    {{site.data.alerts.end}}

    {% include_cached copy-clipboard.html %}
    ~~~shell

>     vault write "cockroach_cluster_ca/roles/${node1name}" \

Author: Thanks!

    allow_bare_domains=true \
    common_name="node" \
    allowed_domains="${node1name},localhost,node" \
    ip_sans="${node1addr},${ex_ip}" \
    ext_key_usage="server_auth,client_auth" \
    max_ttl=1000h # just over a month

    vault write "cockroach_cluster_ca/roles/${node2name}" \
    allow_bare_domains=true \
    common_name="node" \
    allowed_domains="${node2name},localhost,node" \
    ip_sans="${node2addr},${ex_ip}" \
    ext_key_usage="server_auth,client_auth" \
    max_ttl=1000h # just over a month 

    vault write "cockroach_cluster_ca/roles/${node3name}" \
    allow_bare_domains=true \
    common_name="node" \
    allowed_domains="${node3name},localhost,node" \
    ip_sans="${node3addr},${ex_ip}" \
    ext_key_usage="server_auth,client_auth" \
    max_ttl=1000h # just over a month 
    ~~~

    ~~~txt

>     Success! Data written to: cockroach_node_ca/roles/cockroach-cluster-alpha-1bth

Reviewer: Maybe use `${node1name}` and so forth in these success messages to reinforce that they will see one success message for each node?

    echo -e $(cat "${secrets_dir}/$node3name/certs.json" | jq .data.certificate | tr -d '"') > "${secrets_dir}/${node3name}/node.crt"
    ~~~

### Provision each node's [trust store](security-reference/transport-layer-security.html#trust-store)

1. Prepare the certs directory on each node:
    {% include_cached copy-clipboard.html %}
    ~~~shell
    gcloud compute ssh $node1name \
    --command 'rm -rf ~/certs && mkdir ~/certs'

    gcloud compute ssh $node2name \
    --command 'rm -rf ~/certs && mkdir ~/certs'

    gcloud compute ssh $node3name \
    --command 'rm -rf ~/certs && mkdir ~/certs'
    ~~~

1. Use SCP to propagate the CA crt and corresponding key pair to each node
    {% include_cached copy-clipboard.html %}
    ~~~shell

    gcloud compute scp -r "${secrets_dir}/$node1name/" !!!"$node1name/certs"
    gcloud compute scp -r "${secrets_dir}/$node2name/" !!!"$node2name/certs"
    gcloud compute scp -r "${secrets_dir}/$node3name/" !!!"$node3name/certs"
    ~~~

### Create a role and issue a certificate for the client

#### Create a PKI role for the root  Issue a root client certificate
Generate a private key/public certificate pair for the root SQL user.

> This is a very powerful private key, as anyone 

Reviewer: Incomplete?

{% include_cached copy-clipboard.html %}
```shell
vault write "cockroach_cluster_ca/roles/root_client" \
    allowed_domains="root" \
    allow_bare_domains=true \
    common_name="root" \
    ext_key_usage="client_auth" \
    max_ttl=10000h # just over a year
```

```txt
Success! Data written to: cockroach_cluster_ca/roles/root_client
```

{% include_cached copy-clipboard.html %}
~~~shell
vault write "cockroach_cluster_ca/issue/root_client" \
    common_name=root --format=json > "${secrets_dir}/root_client/certs.json"
~~~

### Provision the Client's trust store

### Clean up the CA admin jumpbox

> Be sure to delete the private keys (after they've been copied to the nodes). You may have to respond `y` to over-ride permissions in order to delete the keys.

Reviewer: ```suggestion
Be sure to delete the private keys (after they've been copied to the nodes). You may have to respond `y` to override permissions in order to delete the keys.
```
(or use `sudo`, right?

{% include_cached copy-clipboard.html %}
```shell
rm ${secrets_dir}/${node1name}/node.*
rm ${secrets_dir}/${node2name}/node.*
rm ${secrets_dir}/${node3name}/node.*
rm ${secrets_dir}/clients/*
```
## Node operations

### Prepare CockroachDB on each node

#### Install CockroachDB on each node

Download the CockroachDB executable from `binaries.cockroachdb.com`.

{% include_cached copy-clipboard.html %}
```shell
function install_roach_on_node {
    gcloud compute ssh $1 --command 'if [[ ! -e cockroach-v21.2.4.linux-amd64 ]];
    then
        echo "roach not installed; installing roach"
        sudo curl https://binaries.cockroachdb.com/cockroach-v22.1.0-beta.2.linux-amd64.tgz | tar -xz
        sudo cp -i cockroach-v22.1.0-beta.2.linux-amd64/cockroach /usr/local/bin/
        sudo mkdir -p /usr/local/lib/cockroach
        sudo cp -i cockroach-v22.1.0-beta.2.linux-amd64/lib/libgeos.so /usr/local/lib/cockroach/
        sudo cp -i cockroach-v22.1.0-beta.2.linux-amd64/lib/libgeos_c.so /usr/local/lib/cockroach/
    else

>         echo "roach already installed"

Author: refinement is for follow up.

--background
~script~

chmod +x start_roach.sh
gcloud compute scp ./start_roach.sh $node2name:~

# for node 3
cat <<~script~ > start_roach.sh
cockroach start \
--certs-dir=certs \
--advertise-addr="${node3addr}" \
--join="${node1addr},${node2addr},${node3addr}" \
--cache=.25 \
--max-sql-memory=.25 \
--background
~script~

chmod +x start_roach.sh
gcloud compute scp ./start_roach.sh $node3name:~
```

```shell
start_roach.sh                                                                              100%  156     3.0KB/s   00:00
start_roach.sh                                                                              100%  156     3.0KB/s   00:00
start_roach.sh                                                                              100%  156     3.1KB/s   00:00
```

### Start or reset the cluster

#### Start the nodes

To start the cluster, execute the pre-provisioned start script.

{% include_cached copy-clipboard.html %}
```shell
gcloud beta compute ssh --command './start_roach.sh' "${node1name}"
gcloud beta compute ssh --command './start_roach.sh' "${node2name}"
gcloud beta compute ssh --command './start_roach.sh' "${node3name}"
````

#### Reset the nodes

> To cause an already running cluster to switch to the new credentials, send a `SIGHUP` signal to each node.

Reviewer: ```suggestion
To start using the new credentials on a running cluster, send a `SIGHUP` signal to each node.
```

{% include_cached copy-clipboard.html %}
```shell
gcloud beta compute ssh "$node1name" --command 'pkill -SIGHUP -x cockroach'
gcloud beta compute ssh "$node2name" --command 'pkill -SIGHUP -x cockroach'
gcloud beta compute ssh "$node3name" --command 'pkill -SIGHUP -x cockroach'
```

## Client operations

### Provision the client

> #### Install CockroachDB

Reviewer: This is a lot of stacked headings with no intermediate text.

> Install Cockroach locally [another way](install-cockroachdb-mac.html) or with docker as follows:

Reviewer: ```suggestion
Instead of downloading the binary, you can install CockroachDB using Docker:

~~~shell
docker pull cockroachdb/cockroach:latest
~~~

You can also [install the `cockroach` binary locally on macOS](install-cockroachdb-mac.html).
```
If you apply this, change the tildes above to backticks. I couldn't get it to work with backticks inside backticks.

### Vault secrets

> The solution demonstrated here uses HashiCorp Vault to manage these credentials as secrets. This approach benefit from Vault's generally strong security model and ease of use.

Reviewer: ```suggestion
The solution demonstrated here uses HashiCorp Vault to manage these credentials as secrets. This approach benefits from Vault's generally strong security model and ease of use.
```
